timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimeLockServiceMetadataIntegrationTest.java

```diff
@@ -0,0 +1,281 @@
+    private static final LockDescriptor WATCHED_LOCK_1 =
+            AtlasRowLockDescriptor.of(WATCHED_TABLE_NAME, "lock1".getBytes(StandardCharsets.UTF_8));
+    private static final LockDescriptor WATCHED_LOCK_2 =
+            AtlasRowLockDescriptor.of(WATCHED_TABLE_NAME, "lock2".getBytes(StandardCharsets.UTF_8));
+    private static final LockDescriptor WATCHED_LOCK_3 =
+            AtlasRowLockDescriptor.of(WATCHED_TABLE_NAME, "lock3".getBytes(StandardCharsets.UTF_8));
+    private static final LockDescriptor WATCHED_LOCK_4 =
+            AtlasRowLockDescriptor.of(WATCHED_TABLE_NAME, "lock4".getBytes(StandardCharsets.UTF_8));
+    private static final Map<LockDescriptor, ChangeMetadata> ALL_WATCHED_LOCKS_WITH_METADATA = ImmutableMap.of(
+            WATCHED_LOCK_1,
+            ChangeMetadata.unchanged(),
+            WATCHED_LOCK_2,
+            ChangeMetadata.updated("old".getBytes(StandardCharsets.UTF_8), "new".getBytes(StandardCharsets.UTF_8)),
+            WATCHED_LOCK_3,
+            ChangeMetadata.deleted("deleted".getBytes(StandardCharsets.UTF_8)),
+            WATCHED_LOCK_4,
+            ChangeMetadata.created("created".getBytes(StandardCharsets.UTF_8)));
+    private static final String UNWATCHED_TABLE_NAME = "a-random-unwatched-table";
+    private static final LockDescriptor UNWATCHED_LOCK_1 =
+            AtlasRowLockDescriptor.of(UNWATCHED_TABLE_NAME, "lock1".getBytes(StandardCharsets.UTF_8));
+    private static final IdentifiedLockRequest WATCHED_LOCK_REQUEST_WITH_METADATA =
+            standardRequestWithMetadata(ALL_WATCHED_LOCKS_WITH_METADATA);
+
+    private final LockLog lockLog = new LockLog(new MetricRegistry(), () -> 10000L);
+    private final ScheduledExecutorService scheduledExecutorService = PTExecutors.newSingleThreadScheduledExecutor();
+    private final AsyncLockService asyncLockService =
+            AsyncLockService.createDefault(lockLog, scheduledExecutorService, scheduledExecutorService);
+    private final AsyncTimelockServiceImpl timeLockService =
+            new AsyncTimelockServiceImpl(asyncLockService, new InMemoryTimestampService(), lockLog);
+    private final ConjureStartTransactionsRequest startTransactionsRequestWithInitialVersion =
+            ConjureStartTransactionsRequest.builder()
+                    .requestId(UUID.randomUUID())
+                    .lastKnownVersion(ConjureIdentifiedVersion.of(
+                            asyncLockService.leaderTime().id().id(), 0))
+                    .numTransactions(1)
+                    .requestorId(UUID.randomUUID())
+                    .build();
+
+    @Before
+    public void setupLockServiceWithWatches() {
+        timeLockService.startWatching(
+                LockWatchRequest.of(ImmutableSet.of(LockWatchReferences.entireTable(WATCHED_TABLE_NAME))));
+    }
+
+    @Test
+    public void lockEventContainsMetadataFromRequestIfWatched() {
+        waitForFuture(timeLockService.lock(WATCHED_LOCK_REQUEST_WITH_METADATA));
```

Comment:
A neat trick that may be worth trying: you can use a direct executor instead, which then let's you test your code more sequentially; i.e. you won't need to do the `waitForFuture`

Reply:
I don't think a direct executor will help. A request future can only be resolved by other requests unlocking the lock that the request is trying to acquire. I assumed that Awaitility would poll once at the beginning of the wait period, but apparently they don't. I agree that we should never wait for futures, so I have renamed `waitForFuture` to `assertDone` and removed the awaitility stuff. Maybe it is worth to add a test that verifies that metadata behaves nicely with contending requests (for the implementation in this PR, this is trivially true, but might not be for alternative implementations where metadata is kept inside a separate manager and only retrieved when creating the `LockEvent` object).